### PR TITLE
#1927 write config to a temporary file and then rename it

### LINKED ISSUE #1927: [BUG] Malformed config.json file when aborting crc setup on windows
### General information

  * OS:  Windows
  * Hypervisor: Hyper-V
  * Running CRC on:Baremetal-Server 

## CRC version
```powershell
PS C:\Users\crcqe> crc version
CodeReady Containers version: 1.21.0+785d339 (1.22 pre release)
OpenShift version: 4.6.13 (not embedded in executable)
```

## Host Operating System
```powershell
OS Name:                   Microsoft Windows 10 Pro
OS Version:                10.0.18363 N/A Build 18363
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
```

### Steps to reproduce

  1. Fresh new installation for crc (no previous .crc folder)
  2. Execute crc setup 
  3. Instead of answering the question about consent telemtry Y/N. Abort the execution (ctrl + C)
  4. Execute crc setup, actually execute any crc command will fail

### Expected

crc setup will be execute again and the question about consent telemetry can be properly answered. Or any other valid crc command can be executed.

When this setup command is aborted the config.json file is created but as an empty file (other platforms handle this properly creating a file with valid json: {}). So on windows subsequent commands fails as config.json does not have a valid structure.

### Actual

Got an error:

```powershell
PS C:\jenkins-crcqe\workspace\qe\bundle_baremetal_windows10-brno> crc setup --log-level debug                           
time="2021-02-02T14:46:34+01:00" level=fatal msg="error reading configuration file 'C:\\Users\\crcqe\\.crc\\crc.json': While parsing config: unexpected end of JSON input"
```

### Logs

```powershell
PS C:\jenkins-crcqe\workspace\qe\bundle_baremetal_windows10-brno> crc setup --log-level debug                           DEBU CodeReady Containers version: 1.21.0+785d339
DEBU OpenShift version: 4.6.13 (not embedded in executable)
DEBU Running 'crc setup'
CodeReady Containers is constantly improving and we would like to know more about usage (more details at https://developers.redhat.com/article/tool-data-collection)
Your preference can be changed manually if desired using 'crc config set consent-telemetry <yes/no>'
Would you like to contribute anonymous usage statistics? [y/N]:
PS C:\jenkins-crcqe\workspace\qe\bundle_baremetal_windows10-brno> crc setup --log-level debug                           
time="2021-02-02T14:46:34+01:00" level=fatal msg="error reading configuration file 'C:\\Users\\crcqe\\.crc\\crc.json': While parsing config: unexpected end of JSON input"
PS C:\jenkins-crcqe\workspace\qe\bundle_baremetal_windows10-brno> crc start                                             
time="2021-02-02T14:46:38+01:00" level=fatal msg="error reading configuration file 'C:\\Users\\crcqe\\.crc\\crc.json': While parsing config: unexpected end of JSON input"
PS C:\jenkins-crcqe\workspace\qe\bundle_baremetal_windows10-brno> crc config view                                       
time="2021-02-02T14:46:45+01:00" level=fatal msg="error reading configuration file 'C:\\Users\\crcqe\\.crc\\crc.json': While parsing config: unexpected end of JSON input"
```

### PR DESCRIPTION
**Pull request by guillaumerose**
On Windows, when crc setup asks for enabling telemetry, if the user hits
ctrl+c, the program continues a bit, opens the config file and then
exits.
The file crc.json is left empty, corrupted as it is not a valid json.

The flow is:
1. write default configuration {}
2. get a key
3. abort in the middle of setting a key

The solution here is to write the configuration to a temporary file.
When it's done, the temporary file is renamed to the final destination.
On Linux/macOS, the rename is atomic. On Windows, it's not atomic but
it works a bit better.
The file crc.json is not corrupted anymore but an empty file
crc{random number}.json is created in .crc directory.

This method is also used by libmachine to write the machine
configuration.

Fixes  #1927

### CODE REVIEW

**Review by cfergeau — COMMENTED** [review]
I don't understand when the problematic 'abort in the middle of setting a key' happens? Is this during `c.viper.WriteConfigAs`? Trying to understand how we end up writing invalid json to disk.

File: pkg/crc/config/viper_config.go [hunk 1]
```diff
@@ -75,3 +75,22 @@ func ensureConfigFileExists(file string) error {
...
+	tmpFile, err := ioutil.TempFile(filepath.Dir(configFile), pattern)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = os.Remove(tmpFile.Name())
+	}()
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
```
Comment (cfergeau) [inline pkg/crc/config/viper_config.go:91]: bleh, missing a `viper.WriteConfig..(io.Writer)` :-/

**Review by cfergeau — COMMENTED** [review]
Looks fine, just not clear where the bug occurs exactly, and if this change is a workaround for some other issue, or the best fix we can have.

**Comment by guillaumerose** [comment]
Yes this is in the middle of `c.viper.WriteConfigAs`. 

It seems that when I hit ctrl+c here https://github.com/code-ready/crc/blob/master/cmd/crc/cmd/setup.go#L38, the code `config.Set()` is called and it really exits in  `c.viper.WriteConfigAs`.

Really weird. The file is always empty.

**Comment by cfergeau** [comment]
> Yes this is in the middle of `c.viper.WriteConfigAs`.
> 
> It seems that when I hit ctrl+c here https://github.com/code-ready/crc/blob/master/cmd/crc/cmd/setup.go#L38, the code `config.Set()` is called and it really exits in `c.viper.WriteConfigAs`.
> 
> Really weird. The file is always empty.

Can you mention `c.viper.WriteConfigAs` in the log? `3. abort in the middle of setting a key in c.viper.WriteConfigAs`

**Review by cfergeau — APPROVED** [review]

**Pull request merged**